zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/util/XmlUtil.java

```diff
@@ -7,151 +7,116 @@
-import javax.xml.xpath.XPathExpressionException;
-import javax.xml.xpath.XPathFactory;
-import org.apache.commons.lang3.StringUtils;
+import javax.xml.parsers.SAXParserFactory;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.xml.sax.Attributes;
 import org.xml.sax.SAXException;
+import org.xml.sax.helpers.DefaultHandler;
 
 public class XmlUtil {
 
-  public static final String XPATH_NODE_NAME = "//process[@id='%s']/%s[@id='%s']/@name";
   private static final Logger LOG = LoggerFactory.getLogger(XmlUtil.class);
-
   private final ProcessDefinitionServices processDefinitionServices;
 
   public XmlUtil(final ProcessDefinitionServices processDefinitionServices) {
     this.processDefinitionServices = processDefinitionServices;
   }
 
-  public Map<Long, Map<String, String>> getFlowNodesNames(
-      final List<FlowNodeInstanceEntity> flowNodes) {
-    return getNamesForEntities(
-        flowNodes,
-        FlowNodeInstanceEntity::processDefinitionKey,
-        FlowNodeInstanceEntity::flowNodeId,
-        this::getFlowNodeName);
+  private SAXParserFactory getSAXParserFactory() {
+    final SAXParserFactory saxParserFactory = SAXParserFactory.newInstance();
+    saxParserFactory.setNamespaceAware(true);
+    try {
+      saxParserFactory.setFeature("http://xml.org/sax/features/external-general-entities", false);
+      saxParserFactory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+      saxParserFactory.setFeature(
+          "http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+      return saxParserFactory;
+    } catch (final ParserConfigurationException | SAXException e) {
+      throw new RuntimeException("Error creating SAXParser: " + e.getMessage(), e);
+    }
   }
 
-  public String getFlowNodeName(final FlowNodeInstanceEntity flowNode) {
-    final var processDefinition =
-        processDefinitionServices.getByKey(flowNode.processDefinitionKey());
-    return getFlowNodeName(processDefinition, flowNode);
+  public void extractFlowNodeNames(
+      final Long processDefinitionKey, final BiConsumer<Long, ProcessFlowNode> flowNodeConsumer) {
+    final var processDefinition = processDefinitionServices.getByKey(processDefinitionKey);
+    extractFlowNodeNames(processDefinition, flowNodeConsumer);
   }
 
-  private String getFlowNodeName(
-      final ProcessDefinitionEntity processDefinition, final FlowNodeInstanceEntity flowNode) {
-    final var type = BpmnElementType.valueOf(flowNode.type().name());
-    if (type.getElementTypeName().isEmpty()) {
-      return flowNode.flowNodeId();
+  public void extractFlowNodeNames(
+      final Set<Long> processDefinitionKeys,
+      final BiConsumer<Long, ProcessFlowNode> flowNodeConsumer) {
+    final var keysList = new ArrayList<>(processDefinitionKeys);
+    final var result =
+        processDefinitionServices.search(
+            ProcessDefinitionQuery.of(q -> q.filter(f -> f.processDefinitionKeys(keysList))));
```

Comment:
❓ Could we potentially get a paged result and need to ensure we fetch all items? I believe @houssain-barouni has some experience with this in internal methods, i.e. he did this for fetching authorizations. This might only be available in the search client, so moving this there as proposed earlier might help 🙂

Reply:
Good point!
It looks like default size is 100 so we either need to specify higher max (`keysList.size()`) or query all pages.